### DISCUSSION TITLE
Custom Output Type

### DISCUSSION THREAD

**Discussion by mohamad-jawad** [xmartlabs/stock#28] [Ideas]
Hello,
Thanks for the hard-work that has been done for this package.

I want to suggest to add a way to specify a different output type, because in the app I am working on I have different Types for every layer, for example:

- TaskNetwork: Task type in Network layer
- TaskEntity: Task type in local database layer
- TaskModel: Task type in Repository layer used by UI

It would be very beneficial if I can add mappers to have TaskModel as the output type.

so can this ability be added to the package? or I should I add mapper to the output stream to get my custom type?

**Comment by mohamad-jawad** [comment]
Currently I am using this function to do the conversion

```
extension StockResponseExtensions<T> on StockResponse<T> {
  StockResponse<E> convertTo<E>(E Function(T t) mapper) {
    return this.when(
      onLoading: (origin) => StockResponseLoading(origin),
      onData: (origin, t) => StockResponseData(origin, mapper(t)),
      onError: (origin, error, stacktrace) =>
          StockResponseError(origin, error, stacktrace),
    );
  }
}
```

**Reply by mirland** [reply]
Hi! Thanks for your feedback! 
I think your extension is fine, it will work. You can do something like:

```dart
stock.stream(key)
      .map((response) => response.convertTo(mapper))
```

On the other hand, maybe it would be nice to have a method in the package to change the stock type. I mean, what I'm saying is to change the Stock instance type instead of changing the stream type.
Something like:

```dart
Stock<T> stock = ....;
Stock<E> stockE = stock. convertTo(mapper);
```

However, if we want to do that, we have to make some updates to the package. I will keep on my radar.

**Reply by mohamad-jawad** [reply]
The extension is doing the work 👍

**Reply by mirland** [reply]
I included a new extension in the latest version to facilitate this work. Now You can use the `mapData` extension!  More info in #31

**Reply by mohamad-jawad** [reply]
Many Thanks 🌹
